lib/internal/Magento/Framework/Controller/Result/Json.php

```diff
@@ -59,10 +60,20 @@ public function setJsonData($jsonData)
     }
 
     /**
-     * {@inheritdoc}
+     * @param HttpResponse|ResponseInterface $response
+     * @return $this
      */
     protected function render(ResponseInterface $response)
     {
+        return $this->renderHttpResponse($response);
+    }
+
+    /**
+     * @param HttpResponse $response
+     * @return $this
+     */
+    private function renderHttpResponse(HttpResponse $response)
```

Comment:
This private method call is required to make the implicit dependency on the `HttpResponse` explicit.
If an incompatible `ResponseInterface` implementation is passed to `render()`, undefined method errors will be thrown.
With this explicit upcast, the error message will be clear, stating that a `HttpResponse` is required, but a `ResponseInterface` was provided.  
At least from my experience getting errors with clear guidance on types is much more helpful then just undefined method errors. It also makes the dependency clear to anybody reading the code.